llm/pdf-processing/pdf-processing-post-api.md

```diff
@@ -0,0 +1,206 @@
+
+We first specify what we actually want to receive on the output of our UDF by defining the `DataModel`-based `Chunk` class and defining the output column types:
+
+```python
+# Define the output as a DataModel class
+class Chunk(DataModel):
+    key: str
+    text: str
+    embeddings: list[float]
+```
+
+Now we are ready to define the `process_pdf` function itself. We use Python signatures to specify the input and output. Here, `File` is a class used by DataChain to refer to the original file - a PDF document in our case. On the output we use `Iterator` since our function will produce multiple chunks (and so multiple rows in our DataChain table) per original file.
+
+```python
+def process_pdf(file: File) -> Iterator[Chunk]:
+```
+
+The rest of the function definition as well as the definition of `embedding_encoder` specifies how `unstructured` is used to process each individual PDF file. We are using [`partition_via_api`](https://docs.unstructured.io/api-reference/api-services/partition-via-api) to take advantage of the more advanced `"by_similarity"` chunking strategy only offered in the API. If we are not interested in that, it is also possible to do the same using only the free open source unstructured library. For more detail on how this is done and what other PDF processing options there are you can check out the tutorial in the [unstructured documentation](https://docs.unstructured.io/open-source/core-functionality/overview).
+
+Finally, we want the UDF to produce new rows in our DataChain table and so we have it return the Chunk objects we specified above. Here, we use `yield` instead of `return` as each PDF file produces several Chunk objects.
+
+```python
+    # Add new rows to DataChain
+    for chunk in chunks_embedded:
+        yield Chunk(
+            key=file.path,
+            text=chunk.text,
+            embeddings=chunk.embeddings,
+        )
+```
+
+## DataChain versioning and additional SaaS features
+
+Now, we have a versioned dataset with a new version automatically created by DataChain whenever we run the script above. By default, versions are numbered and we can always recall a particular version by specifying it as in this call:
+
+```python
+DataChain.from_dataset("embedded-documents", version=<version number>).show()
+
+```
+
+In the SaaS version of DataChain we also provide lineage tracking for datasets, more robust dataset versioning and auditability, a managed on-demand compute including with GPU clusters to really scale up your workloads and a graphical user interface among other features.
+
+DataChain SaaS is currently available as a private preview in our [DVC Studio](https://studio.dvc.ai/) platform right now, feel free to ask us about that at `support@iterative.ai`!
```

Comment:
I think the correct link is: https://studio.dvc.ai/auth/sign-up

But I'm honestly not sure and that one will get them there too if they click on sign up.  I don't see a waiting list link - only a sign-up for Studio link that says 'Contact Us' which doesn't even really make sense.